### ISSUE TITLE
What is the magic number mean in HttpBasics.lesson/2 ?

### ISSUE THREAD

**Issue by aiddroid** [WebGoat/WebGoat#494]
What is the magic number mean in HttpBasics.lesson/2.I only saw a "person" send to http://localhost:8080/WebGoat/HttpBasics/attack1, it confused me.

**Comment by aiddroid** [comment]
The 3rd step's  answer depends on itself. I don't think it a good idea.Maybe "magic_number" should send with "person" in step 2.

**Comment by zhangysh1995** [comment]
As the hint says, I suggest you try ZAP or other tool that could read HTTP requests and responses.

**Comment by nbaars** [comment]
@aiddroid did you solve it? Please reopen the issue if you have issues solving it, thanks!

**Comment by itsmeyjc** [comment]
If you capture the request by proxy tool (like zap) then you can see the magic number.
Like this: 
magic_num=12&answer=POST&magic_answer=1

Then enter the magic_num the input field.

**Comment by act1vand0** [comment]
Hello @nbaars and @itsmeyjc, and props on this nice learning tool. I ran into this issue as well, and share @aiddroid 's concerns. I'm using Version: 8.0.0.M21. 

**Step 1:** http://127.0.0.1:8080/WebGoat/start.mvc#lesson/HttpBasics.lesson/0
Just read HTTP basics.

**Step 2:** http://127.0.0.1:8080/WebGoat/start.mvc#lesson/HttpBasics.lesson/1
Enter your name and see how the server responds with the reversed version of it. Use a proxy to see the traffic. You only see the following in the POST request:

_POST /WebGoat/HttpBasics/attack1 HTTP/1.1
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:62.0) Gecko/20100101 Firefox/62.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://127.0.0.1:8080/WebGoat/start.mvc
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
X-Requested-With: XMLHttpRequest
Content-Length: 15
Cookie: JSESSIONID=07EAEABCFAD252D56272222992ABB12C
Connection: close

person=testname_

**Step 3:** http://127.0.0.1:8080/WebGoat/start.mvc#lesson/HttpBasics.lesson/2
The question reads:  _What type of HTTP command **did WebGoat use** for this lesson. A POST or a GET._ It's using the past tense **_did use_** and therefore refers to the previous step (step 2), the one you already tried and captured in your proxy. When you look at the request, there's no magic number (as shown above).

It's not clear that you actually need to try the request on step 3 itself, with a dummy value for the magic number question, since you don't have it yet. It's confusing that you need to fail purposely in order to have a first chance to capture the magic number you need to correctly answer the question. 

I can think of two solutions:
1) Include the magic number in step 2, and leave everything else as is, XOR
2) Rename step 3 so that it's clear you need to try the request in step 3 in order to get the magic number.

I suggest option (1) above. I hope all this make sense, and can serve as a tiny contribution to this cool project. Cheers.

**Comment by x25Net** [comment]
Use site inspector, there's a hidden input field called magic_num ;)

**Comment by Jared-Harrington-Gibbs** [comment]
I'm with @act1vand0 on this issue.
We need to include the magic number in step 2 or give clearer instructions.
As it stands, this lesson is not intuitive and makes webgoat seem broken by design for beginners. The last thing we want to do is turn someone away from learning more.

**Comment by leoram2007** [comment]
> If you capture the request by proxy tool (like zap) then you can see the magic number.
> Like this:
> magic_num=12&answer=POST&magic_answer=1
> 
> Then enter the magic_num the input field.

Yeah agree, the trick here is check back the response when you send the wrong magic num! so there is no issue here.

**Comment by blong206b** [comment]
Spent 2.5 hours learning how to setup zap in kali and configure firefox for zap proxy?  What I do on webgoat doesn't show up in zap, but if i open another tab and browse to say cnn zap picks it up. I've closed and opened webgoat. no luck

**Comment by willkeffer3** [comment]
I agree on this as well..........couldn't get ZAP to see the request to begin with. Finally just inspected the page like @jithuvgeorge, and got the info needed.

**Comment by multiwebinc** [comment]
> I'm with @act1vand0 on this issue.
> We need to include the magic number in step 2 or give clearer instructions.
> As it stands, this lesson is not intuitive and makes webgoat seem broken by design for beginners. The last thing we want to do is turn someone away from learning more.

I was about to throw this thing out until I found this thread.

**Comment by Z-Farook** [comment]
![image](https://user-images.githubusercontent.com/46790696/93774640-c22beb80-fc21-11ea-86c2-1b4131e38817.png)
![image](https://user-images.githubusercontent.com/46790696/93777838-8b57d480-fc25-11ea-88ca-8a238c72a736.png)

See **how and form where** is **coming**.

**Comment by MRX465** [comment]
Go to the Developers tools . And there you can get your own magic number.

**Comment by vfxdev** [comment]
> I'm with @act1vand0 on this issue. We need to include the magic number in step 2 or give clearer instructions. As it stands, this lesson is not intuitive and makes webgoat seem broken by design for beginners. The last thing we want to do is turn someone away from learning more.

I agree. It is not intuitive at all and leads to confusion early on. Could we re-open this issue?

**Comment by cokacode** [comment]
If you use `zap` or like I used `burp suite` you can see the request by intercepting it  like this and you can change the answer value later in request by sending it to repeater 

![image](https://user-images.githubusercontent.com/61355855/212532690-ca70caeb-c388-4d70-b13b-213f9731065d.png)

**Comment by cmacd88** [comment]
If you don't use zap or burp or other intercepting proxies, you could do what I did for this one.
I went into the browser's developer tools, on the debugger tab, and looked for the string "magic".

Found this snippet:

```
// sample custom javascript in the recommended way ...
					    // a namespace has been assigned for it, but you can roll your own if you prefer
					    webgoat.customjs.assignRandomVal = function () {
							var x = Math.floor((Math.random() * 100) + 1);
							document.getElementById("magic_num").value = x;
						};
						webgoat.customjs.assignRandomVal();
```

Noticed it assigns a randomized value to a html entity called magic_num.

Then I went into the console, and ran console.log(magic_num)
In my case that yielded what I needed.

Of course there may be places where we can't get around using an intercepting proxy, but this time we could.

**Comment by nbaars** [comment]
I will have a look seems to cause confusion early on when handle with HTTP basics

**Comment by nbaars** [comment]
@act1vand0 I've made some improvements to the lesson. Also removed the references to ZAP as they are introduced later on and not necessary to solve this lesson. Also adjusted the hints etc.
